sale_coupon_auto_refresh/models/sale_order_line.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        lines = super().create(vals_list)
+        lines.mapped("order_id")._auto_refresh_coupons()
+        return lines
+
+    def write(self, vals):
+        def _snapshot():
+            return {
+                d["id"]: {k: v for k, v in d.items() if k != "id"}
+                for d in self.read(tracked_fields, load="")
+            }
+
+        def _is_updated(rec):
+            return old_data.get(rec.id) != new_data.get(rec.id)
+
+        tracked_fields = self._get_tracked_fields_to_recompute_rewards()
+        old_data = _snapshot()
+        res = super().write(vals)
+        new_data = _snapshot()
+        self.filtered(_is_updated).mapped("order_id")._auto_refresh_coupons()
+        return res
+
+    def unlink(self):
+        orders = self.mapped("order_id")
+        res = super().unlink()
+        orders._auto_refresh_coupons()
+        return res
+
+    @api.model
+    def _get_tracked_fields_to_recompute_rewards(self) -> list:
+        """To avoid performance issues, reward lines will be recomputed only
+        if a given subset of sale.order.line field values has been updated
+        when :meth:`write` is called.
+
+        Hook method, to be overridden for custom behaviours.
+
+        :return: a list of sale.order.line field names which will be tracked.
+        """
+        return [
```

Comment:
I don't like this, as it forces to create a glue module each time you add any extension to the promotion mechanism that depends on other fields, ending up with 3 modules: this one, the extension one, and the glue module for adding to this list. Can you let it unconditionally?

Reply:
@pedrobaeza I can, but the `write()` would become like this:
```
    def write(self, vals):
        orders = self.mapped('order_id')
        res = super().write(vals)
        if not self.env.context.get('skip_auto_refresh_coupons'):
            orders |= self.mapped('order_id')
            orders = orders.with_context(skip_auto_refresh_coupons=True)
            orders._auto_refresh_coupons()
        return res
```
And I'm not entirely convinced this would be the best solution, speaking from the POV of performances...

What do you think?

Reply:
Taking into account that the write happens once you click "Save" on the UI, or through code in rare occasions, I think we can accept the trade-off for having extreme compatibility.